File: simulator/src/java/darkness/simulator/dmx/Frame.java
package darkness.simulator.dmx;

import darkness.simulator.ParseException;

public class Frame {
    public static final int SIZE = 512;
	public static final int TRANSPARENT = 256;

    private final int[] channelValues = new int[SIZE];
	private final boolean supportsTransparency;

	public Frame(boolean supportsTransparency) {
		this.supportsTransparency = supportsTransparency;
	}

    public int getChannelValue(int channel) {
        return channelValues[channel - 1];
    }

    public void setChannelValue(int channel, int value) throws ParseException {
        if (value < 0 || !(value <= 255 || value == TRANSPARENT)) {
            throw new IllegalArgumentException(String.format("Value for channel %d has value %d (expected 0-255%s)",
					channel, value, supportsTransparency ? " or " + TRANSPARENT : ""));
        }
        channelValues[channel - 1] = value;
    }
}

File: simulator/src/kotlin/darkness/simulator/dmx/Frame.kt
package darkness.simulator.dmx

import darkness.simulator.ParseException

class Frame(private val supportsTransparency: Boolean) {
    private val channelValues = IntArray(size)

    fun getChannelValue(channel: Int): Int {
        return channelValues[channel - 1]
    }

    @Throws(ParseException::class)
    fun setChannelValue(channel: Int, value: Int) {
        if (value < 0 || !(value <= 255 || value == transparent)) {
            throw IllegalArgumentException(String.format("Value for channel %d has value %d (expected 0-255%s)",
                    channel, value, if (supportsTransparency) " or $transparent" else ""))
        }
        channelValues[channel - 1] = value
    }

    companion object {
        const val size = 512
        const val transparent = 256
    }
}
